File: src/main/java/recipeapplication/security/SignupService.java
package recipeapplication.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;
import recipeapplication.exception.UsernameExistsException;
import recipeapplication.model.User;
import recipeapplication.repository.UserRepository;

@Service
public class SignupService
{
    private UserRepository userRepository;

    @Autowired
    public SignupService(UserRepository userRepository)
    {
        this.userRepository = userRepository;
    }

    public void registerNewUser(UserDto userDto) throws UsernameExistsException
    {
        if (usernameAlreadyRegistered(userDto.getUsername()))
        {
            throw new UsernameExistsException();
        }

        User user = new User();

        user.setUsername(userDto.getUsername());
        user.setEmail(userDto.getEmail());
        user.setPassword(
                new BCryptPasswordEncoder(11).encode(userDto.getPassword())
        );

        userRepository.save(user);
    }

    private boolean usernameAlreadyRegistered(String username)
    {
        return (userRepository.findByUsername(username) != null);
    }
}

File: src/main/kotlin/thecookingpot/service/SignupService.kt
package thecookingpot.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import thecookingpot.dto.UserDto
import thecookingpot.exception.EmailExistsException
import thecookingpot.exception.UsernameExistsException
import thecookingpot.model.User
import thecookingpot.repository.UserRepository

@Service
class SignupService @Autowired constructor(private val userRepository: UserRepository, private val passwordEncoder: PasswordEncoder) {
    @Throws(UsernameExistsException::class, EmailExistsException::class)
    fun registerNewUser(userDto: UserDto) {
        if (usernameAlreadyRegistered(userDto.username)) {
            throw UsernameExistsException()
        }

        if (emailAlreadyRegistered(userDto.email)) {
            throw EmailExistsException()
        }

        val user = User().apply {
            username = userDto.username
            email = userDto.email
            password = passwordEncoder.encode(userDto.password)
            newUser = true
        }

        userRepository.save(user)
    }

    private fun usernameAlreadyRegistered(username: String): Boolean {
        return userRepository.findByUsername(username).isPresent
    }

    private fun emailAlreadyRegistered(email: String): Boolean {
        return userRepository.findByEmail(email).isPresent
    }
}